packages/api/src/user.js

```diff
@@ -62,16 +89,53 @@ function createMagicLoginController (env, testModeBypass = magicLinkBypassForE2E
   }
 }
 
+/**
+ * @param {UserRegistrationContext} env
+ * @returns {RequestAuthenticator}
+ */
+const createMagicLinkRequestAuthenticator = (env) => async (request) => {
+  const auth = request.headers.get('Authorization') || ''
+  const token = env.magic.utils.parseAuthorizationHeader(auth)
+  const metadata = await (createMagicLoginController(env).authenticate({ token }))
+  /** @type {IssuedAuthentication} */
+  const authentication = {
+    ...metadata,
+    issuer: metadata.issuer,
+    publicAddress: metadata.publicAddress
+  }
+  return authentication
+}
+
+/**
+ * Initialize a new user that just registered.
+ * @param {object} ctx
+ * @param {object} user
+ * @param {string} user.id
+ * @param {string} user.issuer
+ */
+async function initializeNewUser (ctx, user) {
+  await initializeBillingForNewUser(
+    {
+      customers: ctx.customers,
+      subscriptions: ctx.subscriptions,
+      user: { ...user, id: user.id.toString() }
+    }
+  )
+}
+
 /**
  * @param {Request} request
- * @param {import('./env').Env} env
+ * @param {UserRegistrationContext} env
+ * @returns {Promise<{ issuer: string }>}
  */
 async function loginOrRegister (request, env) {
   const data = await request.json()
-  const auth = request.headers.get('Authorization') || ''
 
-  const token = env.magic.utils.parseAuthorizationHeader(auth)
-  const metadata = await (createMagicLoginController(env).authenticate({ token }))
+  const authenticateRequest = env.authenticateRequest || createMagicLinkRequestAuthenticator(env)
```

Comment:
eventually I want to make this only come from `env`, and probably also be a CompositeRequestAuthenticator(...requestAuthenticators) that tries several in order.
* a la https://docs.djangoproject.com/en/4.1/topics/auth/customizing/#specifying-authentication-backends


That way we can support magic link in addition to other ways of authenticating requests with a standard abstraction. Then we can e.g. stop using magic-link-specific stuff in all our tests. (then we can drop magic when/if appropriate)

Reply:
I agree it makes sense to move this to `env` eventually, especially since `createMagicLinkRequestAuthentication` pulls everything it needs from `env` anyway.